controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java

```diff
@@ -190,222 +188,149 @@ public void cancelChange(ApplicationId applicationId, boolean keepApplicationCha
-    private boolean deploymentComplete(LockedApplication application) {
-        return order.jobsFrom(application.deploymentSpec()).stream()
-                    .filter(JobType::isProduction)
-                    .filter(job -> job.zone(controller.system()).isPresent())
-                    .allMatch(job -> changeDeployed(application, job));
+    private ApplicationController applications() {
+        return controller.applications();
     }
 
-    /**
-     * Returns whether the given application should skip deployment of its current change to the given production job zone.
-     *
-     * If the currently deployed application has a newer platform or application version than the application's
-     * current change, the method returns {@code true}, to avoid a downgrade.
-     * Otherwise, it returns whether the current change is redundant, i.e., all its components are already deployed.
-     */
-    private boolean changeDeployed(Application application, JobType job) {
+    /** Returns the instant when the given application's current change was completed for the given job. */
+    private Optional<Instant> changeCompletedAt(Application application, JobType job) {
         if ( ! job.isProduction())
             throw new IllegalArgumentException(job + " is not a production job!");
 
         Deployment deployment = application.deployments().get(job.zone(controller.system()).get());
         if (deployment == null)
-            return false;
+            return Optional.empty();
 
         int applicationComparison = application.change().application()
                                                .map(version -> version.compareTo(deployment.applicationVersion()))
                                                .orElse(0);
 
-        int platformComparion = application.change().platform()
-                                           .map(version -> version.compareTo(deployment.version()))
-                                           .orElse(0);
+        int platformComparison = application.change().platform()
+                                            .map(version -> version.compareTo(deployment.version()))
+                                            .orElse(0);
 
-        if (applicationComparison == -1 || platformComparion == -1)
-            return true; // Avoid downgrades!
+        // TODO jvenstad: Allow downgrades when considering whether to trigger -- stop them at choice of deployment version.
+        // TODO jvenstad: This allows tests to be re-run, for instance, while keeping the deployment itself a no-op.
+        if (applicationComparison == -1 || platformComparison == -1)
+            return Optional.of(deployment.at());
 
-        return applicationComparison == 0 && platformComparion == 0;
+        return applicationComparison == 0 && platformComparison == 0 ? Optional.of(deployment.at()) : Optional.empty();
```

Comment:
Last couple of lines can be simplified to:

```
Optional.of(deployment.at()).filter(ignored -> {
  applicationComparison == -1 || platformComparion == -1 ||  
  (applicationComparison == 0 && platformComparion == 0)
});
```

Reply:
Sure!